File: packages/sdk/go/x402_client.go
package x402

import (
	"github.com/gagliardetto/solana-go"
	"github.com/x402-upl/sdk/go/tap"
)

type X402Client struct {
	wallet        solana.PrivateKey
	solanaClient  *SolanaX402Client
	discovery     *ServiceDiscovery
	tapClient     *tap.TAPClient
	config        X402Config
}

func NewX402Client(wallet solana.PrivateKey, config X402Config) (*X402Client, error) {
	solanaClient, err := NewSolanaX402Client(wallet, config)
	if err != nil {
		return nil, err
	}

	discovery := NewServiceDiscovery(config.RegistryAPIURL)

	var tapClient *tap.TAPClient
	if config.EnableTAP && config.TAPConfig != nil {
		tapClient = tap.NewTAPClient(config.TAPConfig, config.AgentIdentity)
	}

	return &X402Client{
		wallet:       wallet,
		solanaClient: solanaClient,
		discovery:    discovery,
		tapClient:    tapClient,
		config:       config,
	}, nil
}

func (c *X402Client) Discover(options map[string]interface{}) ([]X402Service, error) {
	return c.discovery.Discover(options)
}

func (c *X402Client) GetService(serviceID string) (*X402Service, error) {
	return c.discovery.GetService(serviceID)
}

func (c *X402Client) Get(url string, params map[string]string) (interface{}, error) {
	return c.solanaClient.Get(url, params)
}

func (c *X402Client) Post(url string, data interface{}, params map[string]string) (interface{}, error) {
	return c.solanaClient.Post(url, data, params)
}

func (c *X402Client) RegisterAgent(stake *float64) (*tap.AgentIdentity, error) {
	if c.tapClient == nil {
		return nil, NewValidationError("TAP must be enabled to register as an agent")
	}

	walletAddress := c.wallet.PublicKey().String()
	return c.tapClient.RegisterAgent(walletAddress, stake)
}

func (c *X402Client) DiscoverAgents(filters map[string]string) ([]*tap.AgentIdentity, error) {
	if c.tapClient == nil {
		return nil, NewValidationError("TAP must be enabled to discover agents")
	}

	return c.tapClient.DiscoverAgents(filters)
}

func (c *X402Client) GetAgentIdentity() *tap.AgentIdentity {
	if c.tapClient == nil {
		return nil
	}
	return c.tapClient.GetAgentIdentity()
}

func (c *X402Client) GetTAPClient() *tap.TAPClient {
	return c.tapClient
}

func (c *X402Client) GetWallet() solana.PrivateKey {
	return c.wallet
}

func (c *X402Client) GetWalletAddress() string {
	return c.wallet.PublicKey().String()
}

func (c *X402Client) GetNetwork() string {
	return c.config.Network
}

func (c *X402Client) GetMetrics() PaymentMetrics {
	return c.solanaClient.metrics
}

func (c *X402Client) GetPaymentHistory(limit int) []PaymentRecord {
	if limit <= 0 || limit > len(c.solanaClient.paymentHistory) {
		return c.solanaClient.paymentHistory
	}
	return c.solanaClient.paymentHistory[len(c.solanaClient.paymentHistory)-limit:]
}

File: packages/sdk/python/x402_upl/x402_client.py
from typing import Any, Dict, Optional

from solders.keypair import Keypair

from .client import SolanaX402Client, X402Config
from .service_discovery import ServiceDiscovery
from .tap import TAPClient, TAPConfig, AgentIdentity


class X402Client:
    def __init__(
        self,
        wallet: Keypair,
        config: Optional[X402Config] = None
    ):
        self.wallet = wallet
        self.config = config or X402Config()

        self.solana_client = SolanaX402Client(wallet, self.config)
        self.discovery = ServiceDiscovery(self.config.registry_api_url)

        self.tap_client: Optional[TAPClient] = None
        if self.config.enable_tap and self.config.tap_config:
            self.tap_client = TAPClient(
                self.config.tap_config,
                self.config.agent_identity
            )

    async def discover(self, options: Optional[Dict[str, Any]] = None):
        return await self.discovery.discover(options or {})

    async def get_service(self, service_id: str):
        return await self.discovery.get_service(service_id)

    async def get(self, url: str, params: Optional[Dict[str, Any]] = None) -> Any:
        return await self.solana_client.get(url, params)

    async def post(
        self,
        url: str,
        data: Optional[Dict[str, Any]] = None,
        params: Optional[Dict[str, Any]] = None
    ) -> Any:
        return await self.solana_client.post(url, data, params)

    async def register_agent(self, stake: Optional[float] = None) -> AgentIdentity:
        if not self.tap_client:
            raise ValueError("TAP must be enabled to register as an agent")

        wallet_address = str(self.wallet.pubkey())
        return await self.tap_client.register_agent(wallet_address, stake)

    async def discover_agents(
        self,
        filters: Optional[Dict[str, Any]] = None
    ) -> list[AgentIdentity]:
        if not self.tap_client:
            raise ValueError("TAP must be enabled to discover agents")

        return await self.tap_client.discover_agents(filters)

    def get_agent_identity(self) -> Optional[AgentIdentity]:
        if self.tap_client:
            return self.tap_client.agent_identity
        return None

    def get_tap_client(self) -> Optional[TAPClient]:
        return self.tap_client

    def get_connection(self):
        return self.solana_client.connection

    def get_wallet(self) -> Keypair:
        return self.wallet

    def get_wallet_address(self) -> str:
        return str(self.wallet.pubkey())

    def get_network(self) -> str:
        return self.config.network

    async def get_balance(self, asset: str = "SOL") -> float:
        return await self.solana_client.get_balance(asset)

    def get_metrics(self):
        return self.solana_client.metrics

    def get_payment_history(self, limit: Optional[int] = None):
        if limit:
            return self.solana_client.payment_history[-limit:]
        return self.solana_client.payment_history

    async def close(self):
        await self.solana_client.close()
        if self.tap_client:
            await self.tap_client.close()
